datadog/dogstatsd/base.py

```diff
@@ -312,40 +330,60 @@ def _report(self, metric, metric_type, value, tags, sample_rate):
 
         # Resolve the full tag list
         tags = self._add_constant_tags(tags)
-
-        # Create/format the metric packet
-        payload = "%s%s:%s|%s%s%s" % (
-            (self.namespace + ".") if self.namespace else "",
-            metric,
-            value,
-            metric_type,
-            ("|@" + text(sample_rate)) if sample_rate != 1 else "",
-            ("|#" + ",".join(tags)) if tags else "",
-        )
+        payload = self._serialize_metric(metric, metric_type, value, tags, sample_rate)
 
         # Send it
         self._send(payload)
 
+    def _reset_telementry(self):
+        self.metrics_count = 0
+        self.events_count = 0
+        self.service_checks_count = 0
+        self.bytes_sent = 0
+        self.bytes_dropped = 0
+        self.packets_sent = 0
+        self.packets_dropped = 0
+
+    def _flush_telemetry(self):
+        return [
+            self._serialize_metric("datadog.dogstatsd.client.metrics", "c", self.metrics_count, self.telemetry_tags, 1),
+            self._serialize_metric("datadog.dogstatsd.client.event", "c", self.events_count, self.telemetry_tags, 1),
+            self._serialize_metric("datadog.dogstatsd.client.service_checks", "c", self.service_checks_count, self.telemetry_tags, 1),
+            self._serialize_metric("datadog.dogstatsd.client.bytes_sent", "c", self.bytes_sent, self.telemetry_tags, 1),
+            self._serialize_metric("datadog.dogstatsd.client.bytes_dropped", "c", self.bytes_dropped, self.telemetry_tags, 1),
+            self._serialize_metric("datadog.dogstatsd.client.packets_sent", "c", self.packets_sent, self.telemetry_tags, 1),
+            self._serialize_metric("datadog.dogstatsd.client.packets_dropped", "c", self.packets_dropped, self.telemetry_tags, 1),
+        ]
+
     def _send_to_server(self, packet):
+        success = False
         try:
             # If set, use socket directly
             (self.socket or self.get_socket()).send(packet.encode(self.encoding))
+            self._reset_telementry()
+            success = True
         except socket.timeout:
             # dogstatsd is overflowing, drop the packets (mimicks the UDP behaviour)
-            return
+            pass
         except (socket.error, socket.herror, socket.gaierror) as se:
             log.warning("Error submitting packet: {}, dropping the packet and closing the socket".format(se))
             self.close_socket()
         except Exception as e:
             log.error("Unexpected error: %s", str(e))
-            return
+
+        self.packets_sent += 1
+        self.bytes_sent += len(packet)
+        if not success:
+            self.bytes_dropped += len(packet)
+            self.packets_dropped += 1
 
     def _send_to_buffer(self, packet):
         self.buffer.append(packet)
         if len(self.buffer) >= self.max_buffer_size:
             self._flush_buffer()
 
     def _flush_buffer(self):
+        self.buffer.extend(self._flush_telemetry())
```

Comment:
If not using a buffer, there will be no telemetry since you're only doing this here right ?